starboard/tools/api_leak_detector/api_leak_detector.py

```diff
@@ -85,17 +85,17 @@
 _UNKNOWN_SOURCE_FILES = 'unknown_source_file(s)'
 
 
-def _DiffWithManifest(leaked_symbols, manifest_path):
```

Comment:
Instead of making all the private functions public and directly testing them, I think it would be better to have more behavior-driven tests that target the main function. The benefits of doing this are mostly summarized in go/unit-testing-practices#public-apis and go/unit-testing-practices#behavior-testing.

I'm thinking some example test cases using this pattern could be something like:
* `test_manifest_mode_outputs_leaking_symbols`
* `test_submit_check_mode_leaks_removed_returns_error_code`
* `test_submit_check_mode_leaks_added_returns_error_code`
* `test_submit_check_mode_no_diff_returns_success`

I know this would require some significant changes to the PR, but please lmk what you think.

Reply:
So I like the idea but because api_leak_detector is purely functional and not object oriented I think it needs methods tests as well. The only "public" function currently in api_leak_detector is `main`, which is testing _a lot_ of functionality at a very high level

My preference would be to add these tests (then someone who break the leak detector can more easily find where) then also add higher level tests

Reply:
I guess I don't consider it a bad thing that `main()` is testing a lot of functionality at a high level. To me, the `main()` method and the set of command-line flags define the interface of the tool. And a developer making changes to this tool in the future will want to make sure that that interface is maintained, i.e., there are no unexpected changes to the defined behaviors of the tool.

And go/unit-testing-practices#public-apis seems to share that view: "Tests that are independent of implementation details...Because they test at a higher level, they have the power to expose bugs 'in the seams', that lower-level tests can't. "

Reply:
Yes, I agree that adding high level tests would be good here as well, but I'm saying that low level tests are good too to make it clearer for developers who break the tests what they broke as when we use the `main` entry point to test api leak detector there are a ton of places where things could break.

Also, yes I agree that behaviors should be tested and not methods, as you link. That isn't implying that you shouldn't use methods though, just that you should keep tests focused in testing a single behavior instead of validating multiple behaviors with a single method. I believe this conforms to that

Reply:
One more thing is that `main` is a huge function with little hierarchical structure—it directly calls i.e. `_RunCommand(['ninja', ...])` to build relevant targets, directly calls `nm`, etc., so would have to be completely restructured if we were to use that directly

Reply:
Just to summarize our offline discussion - I'm still of the opinion that the `main()` method and the set of command-line flags together define the interface of the tool and that go/unit-testing-practices#public-apis-examples applies. In particular, that we shouldn't "increase visibility for methods that would otherwise be private in order to expose them to unit tests."

We do agree that the tests as currently written will provide helpful coverage, though, so I'm approving this and will file a ticket to consider rewriting the tests as described in my first comment (and doing any necessary refactoring at that time).